Added Query feature

Humbarrt opened the pull request:
Hi there,

I wanted the clock to be able to utilize the query as input. Since other people were asking for the same feature (#168), I figured I might take a shot at it. My implementation isn't pretty, but it's a start. This is my first time working on a Grafana plugin. 
Have a great day!

Humbarrt committed:
added queries

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/clock-panel?pullRequest=179) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/clock-panel?pullRequest=179) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/clock-panel?pullRequest=179) it.</sub>

academo commented on src/options.ts:
```diff
@@ -66,7 +94,48 @@ function addCountdown(builder: PanelOptionsEditorBuilder<ClockOptions>) {
+      path: 'countdownSettings.queryField',
+      name: 'Field',
+      settings: {
+        options: data?.[0]?.fields?.map((f: any) => ({ label: f.name, value: f.name })) ?? [],
```
You can use instead `addFieldNamePicker` from `grafana-data` (and in the count up)

academo commented on src/types.ts:
```diff
@@ -12,6 +12,31 @@ export interface ClockOptions {
+  query = 'query',
+}
+
+export enum CountdownQueryCalculation {
```
these two types are almost the same. extract the common part. e.g; `QueryCalculation`

academo commented on src/ClockPanel.tsx:
```diff
@@ -66,6 +66,7 @@ export function ClockPanel(props: Props) {
         now={now}
         replaceVariables={props.replaceVariables}
         options={props.options}
+        data={data}
```
you are passing all the data and responsibility of calculate your field to a component that should merely render data. Instead, create another function that abstracts all the logic and returns the field you actually want and then pass it here.

academo replied:
you also want a loading state. since you can only use data when it is done. You will want to show a [spinner](https://developers.grafana.com/ui/latest/index.html?path=/story/visualizations-spinner--basic) while the data loads when the source is a query or it'll show blank while it loads.

Humbarrt replied:
I dont know if you want to show a spinner every time the data loads. Grafana itself shows if a query is done and this logic isn't taking that much more time.

academo replied:
you are right! Let's just leave it as is :)

academo commented on src/components/RenderTime.tsx:
```diff
@@ -1,30 +1,109 @@
+      );
+      break;
+    case ClockSource.query:
+      if (data.state === 'Done' && data.series.length > 0) {
```
what would it display when there are not data series and the state is done? 
There should be feedback for the user indicating that the clock can't display because there's not data.

academo commented on src/components/RenderTime.tsx:
```diff
@@ -1,30 +1,109 @@
+    case ClockSource.query:
+      if (data.state === 'Done' && data.series.length > 0) {
+        let field_name = countdownSettings.queryField;
+        if (
```
nit: you can reduce all this condition to `!field_name?.length || !data?.series?.[0]?.fields?.length`

academo commented on src/components/RenderTime.tsx:
```diff
@@ -1,30 +1,109 @@
+        }
+
+        let value: Moment | undefined = moment();
+        switch (countdownSettings.queryCalculation) {
```
all this logic should be extracted to another function and file.

academo commented on src/components/RenderTime.tsx:
```diff
@@ -1,30 +1,109 @@
+        let value: Moment | undefined = moment();
+        switch (countdownSettings.queryCalculation) {
+          case CountdownQueryCalculation.lastNotNull:
+            value = moment(field_values.filter((v: any) => v !== null && !Number.isNaN(v)).at(-1));
```
please do not use the `any` type

academo commented on src/components/RenderTime.tsx:
```diff
@@ -69,23 +148,94 @@ function getCountdownText({
+        )
+      );
+      break;
+    case ClockSource.query:
```
all this logic is duplicated.  only the return text changes.

academo reviewed (changes requested):
Thanks so much for your contribution!

You've done great for your first time working in a grafana plugin.

There are several things I'd like you to work on before we merge this PR. Most of them go around code orgnanization and de-duplication rather than the functionality which looks good to me.

There's an important part on user feedback when there's no suitable data to show. For example, right now it'll just show the end text in situations where there's no suitable data to render, instead it should tell the user that the data can not be used for a clock. Think about the user experience and what would you like, as an user, to see if there's something wrong in the data you are using.

Give it a try, throw data at it that makes no sense, something an user might just do, and see what happens. Try several data series too.

Humbarrt committed:
refactor query calculation

Humbarrt committed:
Splitting time rendering and computation

Humbarrt committed:
fixed error display

Humbarrt committed:
Added Description

Humbarrt commented:
Hi,
I changed things according to your requests. I added a No Value Text and a Invalid Value Text, so if a user wants to debug their query, they can. In production you might want an empty field, thus I made it changeable. I also added a description under the countdown/countup so that you can have a label for the query result... something like:
`{
'name': ['Next New Year', 'Last New Year', 'Year After Next Year'],
'event': ['2025-01-01', '2024-01-01', '2026-01-01']
}`
where the 'name' could be the description and the 'event' could be the countup/countdown time. The description can also be static or not be displayed at all.

In my tests i found that all normal time inputs are working, except the epoch input. But a user could fix that with a grafana data transform to multiply the row by 1000, so I think it's no big deal. In the future you maybe want to consider a custom input format though. 

Anyway, I know ComputeTime.tsx isn't super pretty, since it handles both time and description, maybe you have a better idea on how to handle that. 
Cheers!

academo commented on src/components/RenderTime.tsx:
```diff
@@ -31,76 +24,36 @@ function getCountdownText({
     case ClockSource.input:
       if (!countdownSettings.endCountdownTime) {
-        return countdownSettings.endText;
+        return countdownSettings.noValueText;
```
even though this makes sense it is best to not change this existing behaviour or people will get unexpected behaviours when they update their plugin.

academo commented on src/ClockPanel.tsx:
```diff
@@ -53,6 +54,17 @@ export function ClockPanel(props: Props) {
   }, [props.options.refresh, timezoneToUse]);
 
+  //refresh the time
+  let [time, description, err]: [moment.Moment, string, string | undefined] = useMemo(() => {
```
I was rather confused with this variable and the `getTime` function. I now understand this `time` is the "targetTime". The time the counter will go up or down. Perhaps you should call it as such.

Also it is ok if you want to get the `description` from this same function but you can do two things:

* move this function out of the ComputeTime.tsx component. Since that is a react component and there's no reason to have it there. move it to its own function and perhaps you can rename it to something such as `calculateClockOptions` or `getProcessedClockOptions`. something that tells this function is doing calculations based on the options it gets.

academo commented on src/options.ts:
```diff
@@ -125,6 +272,81 @@ function addCountup(builder: PanelOptionsEditorBuilder<ClockOptions>) {
+    })
+    .addTextInput({
+      category,
+      path: 'descriptionSettings.description',
```
To avoid this getting confusing let's not call this "description" you can maybe it called `descriptionText` or `descriptionContent` otherwise it is not clear what is inside this in the rest of the code.

academo commented on .config/.eslintrc:
```diff
@@ -15,6 +15,7 @@
       "plugins": ["deprecation"],
       "files": ["src/**/*.{ts,tsx}"],
       "rules": {
+        "no-redeclare": "warn",
```
Is there any particular reason you are changing this to a warning?

Humbarrt replied:
yes, when i changed the types as you requested
> these two types are almost the same. extract the common part. e.g; `QueryCalculation`

I used const/type instead of enum, but I got an error for defining it twice. That's why I changed it to be a warning:
clock-panel/src/types.ts
40:13  warning  'CountdownQueryCalculation' is already defined  no-redeclare
46:13  warning  'CountupQueryCalculation' is already defined    no-redeclare

I agree that I shouldn't have disabled it overall. Now I disable it by line and I've removed the line from [.config/.eslintrc](https://github.com/grafana/clock-panel/pull/179/files/39e5c0d73a36297329823703140727089ec3fd03#diff-71f3678a7bd77030e1c7ecacee33c4b70ba99e6e7f0c1dfe0243ef3e700d6e07)

academo reviewed (changes requested):
Hi @Humbarrt I really like the recent changes to consolidate all the logic to calculate the target time and the description addition. though I would had rather you put the description part in a separate PR to make this review faster, since it is not core to the query functionality. You can keep it here for now since I already reviewed. 

I added some comments, many of them are nits about variable names and code style changes to make the code more legible but the core logic seems good to go once you solve the problem with multiple series. I have also tested the plugin locally and it works just as expected with the backwards compatibility we need with previous versions. 

before we merge it, add unit tests for the cases you created.

 You updated the tests so the types work but you didn't add test cases for displaying the data based on a series of data.  This is important so we know the plugin is working in future updates. I recommend you the following test cases to add:

* Count up from query field
* Count down from query field
* Description with static text
* Description from query field
* Finished count up from query field
* Finished count down from query field
* Error state for count up/down
* Error state for description

Humbarrt committed:
change no-redeclare eslint rule

Humbarrt commented on src/components/ComputeTime.tsx:
```diff
@@ -0,0 +1,149 @@
+          _value = values.length > 0 ? sort_values(values).at(-1) : undefined;
+          break;
+        default:
+          console.error('Invalid query calculation', clockSettings.queryCalculation);
```
I only did that because you do the same in [RenderZone.tsx](https://github.com/grafana/clock-panel/blame/e342be1aa37ea63d42aa6631448b863169550a2b/src/components/RenderZone.tsx#L37) - also as a fallback in a switch statement. My console.log should not happen, it's only triggering if someone adds QueryCalculations without adding cases to the switch.

Humbarrt replied:
I only did that because you did the same in [RenderZone.tsx](https://github.com/grafana/clock-panel/blame/e342be1aa37ea63d42aa6631448b863169550a2b/src/components/RenderZone.tsx#L37) - also as a fallback in a switch statement. My console.log should not happen, it's only triggered if someone adds QueryCalculations without adding cases to the switch.

Humbarrt committed:
targetTime, descriptionText and null as error

Humbarrt committed:
descriptionText, finalValue and sortedValues

Humbarrt committed:
added QueryRow type

Humbarrt commented on src/components/ComputeTime.tsx:
```diff
@@ -0,0 +1,149 @@
+      }
+
+      let descriptionFieldValues = descriptionField?.values ?? new Array(clockField.values.length);
+      let fieldValues: Array<{ time: MomentInput; description: string }> = clockField.values.map((value, index) => {
```
We use the fieldValues for the QueryCalculation.last &  QueryCalculation.first finalValue and it should not be filtered for that. But I can put 
`let values: QueryRow[] = fieldValues
        .filter((v) => v.time !== null && v.time !== undefined && !Number.isNaN(v.time))
        .map((v) => {
          return { time: moment(v.time), description: v.description };
        });`
into a single loop. Good nit :)

Humbarrt committed:
Split up condition and allowed other series as source

Humbarrt committed:
changing filter and map to a single reduce

Humbarrt commented:
Hi, I think I addressed all your requested changes. Please let me know if you find other problems.
I will do the tests next.
Cheers!

academo commented on src/components/CalculateClockOptions.ts:
```diff
@@ -0,0 +1,161 @@
+    case ClockSource.query:
+      const isDataReady = data.state === 'Done' && data.series.length !== 0;
+      const isQueryFieldSelected = clockSettings.queryField && clockSettings.queryField.length !== 0;
+      const isDataSeriesPopulated = data.series && data.series[0].fields && data.series[0].fields.length !== 0;
```
it is possible that we have several series data but some of them came empty so having no fields in series[0] doesn't mean the other don't have data.

Humbarrt replied:
Ok, I will remove the `isDataSeriesPopulated` condition as empty fields would be catched by the `timeField` not finding anything.

academo commented:
> Hi, I think I addressed all your requested changes. Please let me know if you find new problems. I will do the tests next. Cheers!

This looks almost ready to do. looking forward for the test and merging this :D

Humbarrt committed:
removing isDataSeriesPopulated

Humbarrt committed:
added DescriptionRenderer

academo commented:
> Hi, while creating the tests I stumbled upon something. The showDate setting renders the current date and not the targetTime. When I do a countdown/countup it would be useful to know what date I am counting down/up to. This would be more intuitive, at least for me. Now, I know this is only how I feel and changing it would alter existing behavior for users. I just wanted to bring it up. Cheers!

It is best not to change any of the existing behaviour even if that behaviour is somewhat counter intuitive because there are many plugin users that rely on it. Clock panel is rather a popular plugin in Grafana and we don't want to break the workflow of the existing users.

Humbarrt committed:
tests and bug fixes

Humbarrt commented:
Ok, I thought so. As I said, I just wanted to bring it up. Maybe we can add an option to the ShowDate later. 

I've done the tests you asked for. Maybe I'll add some more tomorrow. E.g:
Time/Description from different Series
Null as input
Undefined as input
NaN as input
different modes: lastNotNull, last, firstNotNull, first, min, minFuture, max and maxPast

Cheers!

Humbarrt committed:
more query tests

Humbarrt committed:
all QueryCalculation tests

Humbarrt commented:
Hi, I've done all tests I can think of. If there is anything else I can do, let me know :)
Cheers!

tolzhabayev commented on src/plugin.json:
```diff
@@ -17,18 +14,18 @@
       "large": "img/clock.svg"
     },
     "links": [
-      {"name": "Project site", "url": "https://github.com/grafana/clock-panel"},
```
Please avoid formatting changes that are not relevant to the code change. In this case it would be everything besides `skipDataQuery`

tolzhabayev commented on src/ClockPanel.tsx:
```diff
@@ -2,20 +2,22 @@ import { css } from '@emotion/css';
 import { RenderTime } from 'components/RenderTime';
 import { RenderZone } from 'components/RenderZone';
-import { Moment } from 'moment-timezone';
+import moment, { Moment } from 'moment-timezone';
```
What is the purpose of changing this import?

On the line 59 you can just use `Moment`

Humbarrt replied:
Yes, you are right!

tolzhabayev commented:
The description settings are confusing to the users as they are only relevant for `countdown` and `countup` but are still displayed for any mode.

tolzhabayev commented:
Generally we should add this case to the testing dashboard https://github.com/grafana/clock-panel/tree/main/provisioning/dashboards - maybe together with a testdata datasource to showcase different behaviour.

Humbarrt committed:
updating Moment import

Humbarrt committed:
description bug fix and undi json formatting

Humbarrt commented:
> The description settings are confusing to the users as they are only relevant for `countdown` and `countup` but are still displayed for any mode. The setting should only be visible when relevant.

Imo the description should be available in all modes. But you were right, there was a bug that showed `query` as a description source for the time mode. I fixed that. Now you can have none/input description source for time-mode, countup/countdown without query and none/input/query if you use a query for you countup/countdown

Humbarrt committed:
added panels to testing dashboard and buf fixes

Humbarrt commented:
> Generally we should add this case to the testing dashboard https://github.com/grafana/clock-panel/tree/main/provisioning/dashboards - maybe together with a testdata datasource to showcase different behaviour.

Hi, I added panels to the testing dashboard and fixed a bug that I found in when using the older grafana version.
Cheers!

tolzhabayev reviewed (approved):
@Humbarrt thanks for making the changes and for contributing!

academo reviewed (approved):
@Humbarrt thanks for such a great contribution!

I will merge and release this next week.
